### ISSUE TITLE
Dataset.iterbatches not iterating in batch_size

### ISSUE THREAD

**Issue by proteneer** [deepchem/deepchem#881]
Guys, please stop using linspace unless you know exactly how it behaves. It doesn't do what you think it does and only serves to break things.

``` python
          interval_points = np.linspace(
              0,
              n_samples,
              np.ceil(float(n_samples) / shard_batch_size) + 1,
              dtype=int)

          print("N_SAMPLES", n_samples)
          print("ITVAL POINTS", interval_points[0], interval_points[1], interval_points[2], interval_points[-3], interval_points[-2], interval_points[-1])

          for p_idx, p in enumerate(interval_points):
            if p_idx > 0:
              diff = interval_points[p_idx]-interval_points[p_idx-1]
              print(diff)
              if diff != shard_batch_size:
                print("!!!!!")
```

```
192
192
192
192
191
!!!!!
192
192
192
192
192
192
192
192
191
!!!!!
192
192
192
192
```

**Comment by rbharath** [comment]
It sounds like we should have a unit test on `Dataset.iterbatches` checking that iterates happen in batch size if that's a guaranteed invariant

**Comment by proteneer** [comment]
I fixed this in my ANI PR 19b78462da7e1c60fc7ade6796faf48f3ad73d22 but someone may want to cherry pick the change and apply it to the NumpyDataset as well for the 1.3 release

I'll be out of commission for 2 weeks starting tomorrow.

**Comment by lilleswing** [comment]
We can close this.
